File: ports/rust/src/core/global/bcspointer.rs
use std::rc::Rc;
use std::sync::Arc;

// BcsSharedPointer maps to Rust's Arc for thread-safe shared ownership
pub type BcsSharedPointer<T> = Arc<T>;

// BcsUniquePointer maps to Rust's Box for exclusive ownership
pub type BcsUniquePointer<T> = Box<T>;

// Local (non-thread-safe) shared pointer alias
pub type BcsLocalSharedPointer<T> = Rc<T>;

pub fn new_shared<T>(val: T) -> BcsSharedPointer<T> {
    Arc::new(val)
}

pub fn new_unique<T>(val: T) -> BcsUniquePointer<T> {
    Box::new(val)
}

File: ports/go/src/core/global/bcspointer.go
package global

// In Go, we use native pointers and garbage collection,
// but we define semantic aliases to match the BCS C++ API.

// BcsSharedPointer maps to a native pointer in Go (GC handles sharing)
type BcsSharedPointer[T any] *T

// BcsUniquePointer maps to a native pointer but implies exclusive ownership contextually
type BcsUniquePointer[T any] *T

// NewShared creates a new BcsSharedPointer
func NewShared[T any](val T) BcsSharedPointer[T] {
	return &val
}

// NewUnique creates a new BcsUniquePointer
func NewUnique[T any](val T) BcsUniquePointer[T] {
	return &val
}
